File: android/sleepwalker-core/src/main/kotlin/io/sleepwalker/core/protocol/Status.kt
package io.sleepwalker.core.protocol

/**
 * ACK/status values for the sleepwalker HID protocol.
 *
 * Mirror of firmware status.h and protocol/src/sleepwalker_protocol/status.py.
 * Status notifications are correlated by sequence identifier.
 */
object Status {
    const val RECEIVED: Int = 0x01
    const val QUEUED: Int = 0x02
    const val SENT_TO_USB: Int = 0x03
    const val MALFORMED: Int = 0x10
    const val BAD_CRC: Int = 0x11
    const val DISARMED: Int = 0x12
    const val QUEUE_FULL: Int = 0x13
    const val USB_NOT_MOUNTED: Int = 0x14
    const val UNSUPPORTED_OPCODE: Int = 0x15
    const val KILLED: Int = 0x16

    val ALL: Set<Int> = setOf(
        RECEIVED, QUEUED, SENT_TO_USB, MALFORMED, BAD_CRC, DISARMED,
        QUEUE_FULL, USB_NOT_MOUNTED, UNSUPPORTED_OPCODE, KILLED,
    )

    fun isKnown(status: Int): Boolean = status in ALL

    fun name(status: Int): String = when (status) {
        RECEIVED -> "received"
        QUEUED -> "queued"
        SENT_TO_USB -> "sent_to_usb"
        MALFORMED -> "malformed"
        BAD_CRC -> "bad_crc"
        DISARMED -> "disarmed"
        QUEUE_FULL -> "queue_full"
        USB_NOT_MOUNTED -> "usb_not_mounted"
        UNSUPPORTED_OPCODE -> "unsupported_opcode"
        KILLED -> "killed"
        else -> "unknown"
    }
}

File: protocol/src/sleepwalker_protocol/status.py
"""ACK/status values for the sleepwalker HID protocol.

Status notifications are correlated by sequence identifier (see frame.py)
and emitted on the BLE TX characteristic as structured events. They
distinguish at least: received, queued, sent-to-USB, malformed frame,
bad CRC, disarmed, queue full, USB not mounted, and unsupported opcode.

Status values are uint8 to keep status notifications tiny.
"""
from __future__ import annotations

#: Frame received and parsed successfully. The first positive status in
#: the command lifecycle chain.
STATUS_RECEIVED: int = 0x01

#: Command accepted and enqueued into the hid_bridge queue.
STATUS_QUEUED: int = 0x02

#: Command emitted as a TinyUSB HID report (key-down or key-up report sent).
STATUS_SENT_TO_USB: int = 0x03

#: Frame could not be parsed (bad length, unsupported version, etc).
#: No HID report is emitted.
STATUS_MALFORMED: int = 0x10

#: Frame CRC-32 mismatch. No HID report is emitted.
STATUS_BAD_CRC: int = 0x11

#: Command rejected because firmware safety state is DISARMED.
#: No HID report is emitted.
STATUS_DISARMED: int = 0x12

#: Command rejected because the hid_bridge queue is full.
#: No HID report is emitted.
STATUS_QUEUE_FULL: int = 0x13

#: Command rejected because TinyUSB is not mounted / not ready.
#: No HID report is emitted.
STATUS_USB_NOT_MOUNTED: int = 0x14

#: Command rejected because the opcode is unknown/unsupported.
#: No HID report is emitted.
STATUS_UNSUPPORTED_OPCODE: int = 0x15

#: Command rejected because the safety state is KILLED.
#: No HID report is emitted.
STATUS_KILLED: int = 0x16

#: Set of all known status values for validation.
ALL_STATUSES: frozenset[int] = frozenset({
    STATUS_RECEIVED,
    STATUS_QUEUED,
    STATUS_SENT_TO_USB,
    STATUS_MALFORMED,
    STATUS_BAD_CRC,
    STATUS_DISARMED,
    STATUS_QUEUE_FULL,
    STATUS_USB_NOT_MOUNTED,
    STATUS_UNSUPPORTED_OPCODE,
    STATUS_KILLED,
})


def is_known_status(status: int) -> bool:
    """Return True if the status value is recognized."""
    return status in ALL_STATUSES


__all__ = [
    "STATUS_RECEIVED",
    "STATUS_QUEUED",
    "STATUS_SENT_TO_USB",
    "STATUS_MALFORMED",
    "STATUS_BAD_CRC",
    "STATUS_DISARMED",
    "STATUS_QUEUE_FULL",
    "STATUS_USB_NOT_MOUNTED",
    "STATUS_UNSUPPORTED_OPCODE",
    "STATUS_KILLED",
    "ALL_STATUSES",
    "is_known_status",
]